core/offchain.go

```diff
@@ -214,36 +215,40 @@ func (bc *BlockChain) CommitOffChainData(
 			); err != nil {
 				return NonStatTy, err
 			}
-
+			tempValidatorStats := map[common.Address]*types2.ValidatorStats{}
 			for _, paid := range [...][]reward.Payout{
 				roundResult.BeaconchainAward, roundResult.ShardChainAward,
 			} {
 				for i := range paid {
-					if stats, err := bc.ReadValidatorStats(paid[i].Addr); err == nil {
-						doUpdate := false
-						for j := range stats.MetricsPerShard {
-							if stats.MetricsPerShard[j].Vote.Identity == paid[i].EarningKey {
-								doUpdate = true
-								stats.MetricsPerShard[j].Earned.Add(
-									stats.MetricsPerShard[j].Earned,
-									paid[i].NewlyEarned,
-								)
-							}
+					stats, ok := tempValidatorStats[paid[i].Addr]
+					if !ok {
+						stats, err = bc.ReadValidatorStats(paid[i].Addr)
+						if err != nil {
+							utils.Logger().Info().Err(err).
+								Str("bls-earning-key", paid[i].EarningKey.Hex()).
+								Msg("could not read validator stats to update for earning per key")
+							continue
 						}
-						if doUpdate {
-							if err := rawdb.WriteValidatorStats(
-								batch, paid[i].Addr, stats,
-							); err != nil {
-								utils.Logger().Info().Err(err).
-									Str("bls-earning-key", paid[i].EarningKey.Hex()).
-									Msg("could not update earning per key in stats")
-							}
+						tempValidatorStats[paid[i].Addr] = stats
+					}
+					for j := range stats.MetricsPerShard {
+						if stats.MetricsPerShard[j].Vote.Identity == paid[i].EarningKey {
+							stats.MetricsPerShard[j].Earned.Add(
+								stats.MetricsPerShard[j].Earned,
+								paid[i].NewlyEarned,
+							)
 						}
-					} else {
-						utils.Logger().Info().Err(err).
-							Str("bls-earning-key", paid[i].EarningKey.Hex()).
-							Msg("could not read validator stats to update for earning per key")
 					}
+
+				}
+			}
+			for addr, stats := range tempValidatorStats {
```

Comment:
I know that order doesn't matter here in principle but I want to always enforce reproduciblity so please make a slice of vlaidaotr slice, map to slice, sort.SliceStable the slice, then loop over and write